Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs

```diff
@@ -1,46 +1,276 @@
+        /// </summary>
+        private void DisableAnalyzer(EntityUid uid, GasAnalyzerComponent? component = null, EntityUid? user = null)
+        {
+            if (!Resolve(uid, ref component))
+                return;
+
+            if (user != null && TryComp<ActorComponent>(user, out var actor))
+                _userInterface.TryClose(uid, GasAnalyzerUiKey.Key, actor.PlayerSession);
+
+            if(user is { } userId && component.Enabled)
+                _popup.PopupEntity(Loc.GetString("gas-analyzer-shutoff"), userId, Filter.Entities(userId));
+            component.Enabled = false;
+            SetAppearance(component);
+            Dirty(component);
+            _activeAnalyzers.Remove(uid);
+        }
+
+        /// <summary>
+        /// Disables the analyzer when the user closes the UI
+        /// </summary>
+        private void OnDisabledMessage(EntityUid uid, GasAnalyzerComponent component, GasAnalyzerDisableMessage message)
+        {
+            if (message.Session.AttachedEntity is not {Valid: true} user)
+                return;
+            // The UI should've already been closed but no harm in making sure?
+            DisableAnalyzer(uid, component, user);
+        }
+
+        private void OpenUserInterface(EntityUid user, GasAnalyzerComponent component)
+        {
+            if (!TryComp<ActorComponent>(user, out var actor))
+                return;
+
+            _userInterface.TryOpen(component.Owner, GasAnalyzerUiKey.Key, actor.PlayerSession);
+        }
+
+        /// <summary>
+        /// Fetches fresh data for the analyzer. Should only be called by Update or when the user requests an update via refresh button
+        /// </summary>
+        private bool UpdateAnalyzer(EntityUid uid, GasAnalyzerComponent? component = null)
+        {
+            if (!Resolve(uid, ref component))
+                return false;
+
+            // check if the user has walked away from what they scanned
+            var userPos = EntityManager.GetComponent<TransformComponent>(component.User).Coordinates;
+            if (component.LastPosition.HasValue)
+            {
+                // Check if position is out of range => don't update and disable
+                if (!component.LastPosition.Value.InRange(EntityManager, userPos, SharedInteractionSystem.InteractionRange))
+                {
+                    DisableAnalyzer(uid, component, component.User);
+                    return false;
+                }
+            }
+
+            var gasMixList = new List<GasMixEntry>();
 
-            if (TryComp(args.User, out ActorComponent? actor))
+            if (component.Target != null)
             {
-                component.OpenInterface(actor.PlayerSession, args.ClickLocation);
+                // gas analyzed was used on an entity, try to request gas data
+                var ev = new GasAnalyzerScanEvent();
+                RaiseLocalEvent((EntityUid)component.Target, ev, false);
```

Comment:
```suggestion
                RaiseLocalEvent(component.Target.Value, ev, false);
```

value-type nullability, more obvious what this is doing this way